lib/src/database/matrix_sdk_database.dart

```diff
@@ -0,0 +1,1590 @@
+
+  @override
+  Future<void> deleteFromToDeviceQueue(int id) async {
+    await _toDeviceQueueBox.delete(id.toString());
+    return;
+  }
+
+  @override
+  Future<void> deleteOldFiles(int savedAt) async {
+    final dir = fileStoragePath;
+    final deleteFilesAfterDuration = this.deleteFilesAfterDuration;
+    if (!supportsFileStoring ||
+        dir == null ||
+        deleteFilesAfterDuration == null) {
+      return;
+    }
+    final entities = await dir.list().toList();
+    for (final file in entities) {
+      if (file is! File) continue;
+      final stat = await file.stat();
+      if (DateTime.now().difference(stat.modified) > deleteFilesAfterDuration) {
+        Logs().v('Delete old file', file.path);
+        await file.delete();
+      }
+    }
+  }
+
+  @override
+  Future<void> forgetRoom(String roomId) => transaction(() async {
+        await _timelineFragmentsBox.delete(TupleKey(roomId, '').toString());
+        final eventsBoxKeys = await _eventsBox.getAllKeys();
+        for (final key in eventsBoxKeys) {
+          final multiKey = TupleKey.fromString(key);
+          if (multiKey.parts.first != roomId) continue;
+          await _eventsBox.delete(key);
+        }
+        final preloadRoomStateBoxKeys = await _preloadRoomStateBox.getAllKeys();
+        for (final key in preloadRoomStateBoxKeys) {
+          final multiKey = TupleKey.fromString(key);
+          if (multiKey.parts.first != roomId) continue;
+          await _preloadRoomStateBox.delete(key);
+        }
+        final nonPreloadRoomStateBoxKeys =
+            await _nonPreloadRoomStateBox.getAllKeys();
+        for (final key in nonPreloadRoomStateBoxKeys) {
+          final multiKey = TupleKey.fromString(key);
+          if (multiKey.parts.first != roomId) continue;
+          await _nonPreloadRoomStateBox.delete(key);
+        }
+        final roomMembersBoxKeys = await _roomMembersBox.getAllKeys();
+        for (final key in roomMembersBoxKeys) {
+          final multiKey = TupleKey.fromString(key);
+          if (multiKey.parts.first != roomId) continue;
+          await _roomMembersBox.delete(key);
+        }
+        final roomAccountDataBoxKeys = await _roomAccountDataBox.getAllKeys();
+        for (final key in roomAccountDataBoxKeys) {
+          final multiKey = TupleKey.fromString(key);
+          if (multiKey.parts.first != roomId) continue;
+          await _roomAccountDataBox.delete(key);
+        }
+        await _roomsBox.delete(roomId);
+      });
+
+  @override
+  Future<Map<String, BasicEvent>> getAccountData() =>
+      runBenchmarked<Map<String, BasicEvent>>('Get all account data from store',
+          () async {
+        final accountData = <String, BasicEvent>{};
+        final raws = await _accountDataBox.getAllValues();
+        for (final entry in raws.entries) {
+          accountData[entry.key] = BasicEvent(
+            type: entry.key,
+            content: copyMap(entry.value),
```

Comment:
Similar with previously. Here some refactoring with better types can improve the database at all but as this solution is only a drop in replacement for Hive, I thought this was out of scope

Reply:
However, copyMap is very expensive. I changed it to `makeJson()` which now just does a type cast

Reply:
Is a type cast actually what we want? Iirc the copyMap was used here to prevent anyone from later modifying the cached database values, when they modify the map (because of darts reference semantics), wasn't it?

Reply:
it was because the map was readonly when getting it from other databases. But yes, it could modify the cache in specific circumstances. However, if we copy all Maps and Lists before returning them from the cache, this makes it probably slow. We only need deep copies in some places and not in others.

I changed it now in a way that the underlying key/value stores only return unmodifiable lists and maps so that we are safe here and we do only create a copy now where we actually want to copy them and do only a type cast on the other places. This way we should be always 100% sure that the cache is never corrupted and we should safe some performance as we only copy when it is necessary.